enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,885 @@
+equivalent of `crictl inspect` instead.
+
+#### MCO should learn to pre-load and pin images
+
+Starting with version 4.14 of OpenShift CRI-O will have the capability to pin
+certain images (see [this](https://github.com/cri-o/cri-o/pull/6862) pull
+request for details). That capability will be used to pin all the images
+required for the upgrade, so that they aren't garbage collected by kubelet and
+CRI-O.
+
+Note that pinning images means that kubelet and CRI-O will not remove them, even
+if they aren't in use. It is very important to make sure that there is enough
+available space for these images, as otherwise the performance of the node may
+degrade and it may stop functioning correctly if it runs out of space. The space
+should be enough to accommodate tho releases (current running + candidate for
+install) as well as workload images and buffer.
+
+The changes to pin the images will be done in a
+`/etc/crio/crio.conf.d/pin-upgrade.conf` file, something like this:
+
+```toml
+pinned_images=[
+  "quay.io/openshift-release-dev/ocp-release@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  ...
+]
+```
+
+The images need to be pre-loaded and the CRI-O service needs to be reloaded
+when this configuration changes. To support that a new field will be added to
+the `ContainerRuntimeConfig` object:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: ContainerRuntimeConfig
+metadata:
+  name: pin-upgrade
+spec:
+  containerRuntimeConfig:
+    pinnedImages:
+    - quay.io/openshift-release-dev/ocp-release@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    ...
+```
+
+When the new `pinnedImages` field is added or changed the machine config
+operator will need to pre-load those images (with the equivalent of `crictl
+pull`), create or update the corresponding
+`/etc/crio/crio.conf.d/pin-upgrade.conf` file and ask CRI-O reload its
+configuration (with the equivalent of `systemctl reload crio.service`).
+
+If the `spec.containerRuntimeConfig.imagesDirectory` field is used then it
+should contain a dump of a
+[docker/distribution](https://github.com/distribution/distribution) registry
+server. The machine config operator will first check if there is enough space
+available in the node to copy the images to the `/var/lib/containers/storage`
+directory. Images in the dump of a registry server are compressed, but in
+`/var/lib/containers/storage` they are not, so the machine config operator will
+check that there is at least twice the space space used by the registry server
+dump. If there is not enough space it will be reported as an error conditions in
+the `ContainerRuntimeConfig` object.
+
+If there is enough space then the machine config operator will start a
+temporary embedded image registry server in each node of the cluster, listening
```

Comment:
I think MCO doesn't currently embed an image registry. In the approach that we are suggesting MCO itself doesn't need to embed a registry, only the `bundle-loader` component needs. It is an implementation detail, but I think that `bundle-loader` should live in the MCO repository. So yes, MCO would need to be changed to embed a registry server. This is how we did it in the prototype:

https://github.com/jhernand/upgrade-tool/blob/main/internal/registry.go

Reply:
No, the MCO doesn't currently embed a registry . We did talk about embedding a small one [when we were working on in-cluster builds](https://issues.redhat.com/browse/MCO-571) as a place to store our images, and it would be nice to have if we could also use it for that, but we didn't have the cycles at the time to make it highly available (we'd need to keep instances in sync across the control plane) or coordinate its presence during bootstrap. 

Is there a reason why you want to spin up that image registry on *every* node in the cluster, rather than just spin up a central instance/a deployment spanning some nodes and have CRI-O on the nodes pull from that to cache/pin? 

I mean, ideally for the MCO this would probably be at least two decoupled "general case" components -- e.g.:
1.  "Thing that serves up a small registry instance that images can be batch loaded into (either from a bundle or otherwise)" which the MCO might also be able to use for caching built/layered OS images 
2. "Thing that pins a list of images (from any source) to CRI-O on some list of nodes" which we could use in any case where connectivity between that small registry (or any registry) and the node is in doubt?